sonar-scanner-protocol/src/main/protobuf/scanner_report.proto

```diff
@@ -35,82 +35,117 @@ option java_package = "org.sonar.scanner.protocol.output";
 option optimize_for = SPEED;
 
 message Metadata {
-  optional int64 analysis_date = 1;
+  int64 analysis_date = 1;
   // TODO should we keep this project_key here or not ? Because it's a duplication of Component.key
-  optional string project_key = 2;
-  optional string branch = 3;
-  optional int32 root_component_ref = 4;
-  optional bool cross_project_duplication_activated = 5;
+  string project_key = 2;
+  string branch = 3;
+  int32 root_component_ref = 4;
+  bool cross_project_duplication_activated = 5;
 }
 
 message ActiveRule {
-  optional string rule_repository = 1;
-  optional string rule_key = 2;
-  optional Severity severity = 3;
-  repeated ActiveRuleParam param = 4;
-
-  // TODO replace by map
-  message ActiveRuleParam {
-    optional string key = 1;
-    optional string value = 2;
-  }
+  string rule_repository = 1;
+  string rule_key = 2;
+  Severity severity = 3;
+  map<string,string> params_by_key = 4;
 }
 
 message ComponentLink {
-  optional ComponentLinkType type = 1;
-  optional string href = 2;
+  ComponentLinkType type = 1;
+  string href = 2;
+  
+  enum ComponentLinkType {
+    UNKNOWN_TYPE = 0;
+	  HOME = 1;
+	  SCM = 2;
+	  SCM_DEV = 3;
+	  ISSUE = 4;
+	  CI = 5;
+	}
 }
 
 message Component {
-  optional int32 ref = 1;
-  optional string path = 2;
-  optional string name = 3;
-  optional ComponentType type = 4;
-  optional bool is_test = 5;
-  optional string language = 6;
+  int32 ref = 1;
+  string path = 2;
+  string name = 3;
+  ComponentType type = 4;
+  bool is_test = 5;
+  string language = 6;
   repeated int32 child_ref = 7 [packed = true];
   repeated ComponentLink link = 8;
   // Only available on PROJECT and MODULE types
-  optional string version = 9;
+  string version = 9;
   // Only available on PROJECT and MODULE types
   // TODO rename this property -> batchKey ? moduleKey ?
-  optional string key = 10;
+  string key = 10;
   // Only available on FILE type
-  optional int32 lines = 11;
+  int32 lines = 11;
   // Only available on PROJECT and MODULE types
-  optional string description = 12;
+  string description = 12;
+  
+	enum ComponentType {
+	  UNKNOWN = 0;
+	  PROJECT = 1;
+	  MODULE = 2;
+	  DIRECTORY = 3;
+	  FILE = 4;
+	}
 }
 
 message Measure {
-  optional MeasureValueType value_type = 1;
-  // all values may be unset for "new_xxx" measures (only variations are populated)
-  optional bool boolean_value = 2;
-  optional int32 int_value = 3;
-  optional int64 long_value = 4;
-  optional double double_value = 5;
-  optional string string_value = 6;
-  optional string metric_key = 7;
+  string metric_key = 1;
+  oneof value {
+	  BoolValue boolean_value = 2;
+	  IntValue int_value = 3;
+	  LongValue long_value = 4;
+	  DoubleValue double_value = 5;
+	  StringValue string_value = 6;
+  }
+  
+  message BoolValue {
+    bool value = 1;
+    string data = 2;
```

Comment:
ok but it prevents from moving these primitive objects to constants.proto. Moreover adding a string field "data" to Measure would be more clear IMO.